### PR TITLE
[ObjC] Check entire chain of superclasses to see if class layout is statically known

### LINKED ISSUE #81369: Subclasses of subclasses of NSObject are part of the non-fragile abi even when direct subclasses of NSObject are not
Even when the offsets are fixed, clang is assuming classes that can be proven to inherit from classes that inherit from NSObject and so on are not fixed-width, even though they are and a lot of missed optimizations are because of it. 

@"OBJC_IVAR_$_StaticLayoutSubClass.static_layout_ivar2" = hidden global i64 24

it should be hidden constant i64 24

@interface NSObject {
  int these, will, never, change, ever;
}
@end

@interface StaticLayout : NSObject
@end

@implementation StaticLayout {
  int static_layout_ivar;
}
-(void)meth {
  static_layout_ivar = 0;
  // CHECK-NOT: load i64, ptr @"OBJC_IVAR_$_StaticLayout
}
@end

@interface StaticLayoutSubClass : StaticLayout
@end

@implementation StaticLayoutSubClass {
  int static_layout_ivar2;
}
-(void)meth2 {
  static_layout_ivar2 = 0;
  // CHECK: load i64, ptr @"OBJC_IVAR_$_StaticLayoutSubClass
  // CHECK-NOT: load i64, ptr @"OBJC_IVAR_$_StaticLayoutSubClass
}
@end

### PR DESCRIPTION
**Pull request by AZero13**
As of now, we only check if a class directly inherits from NSObject to determine if said class has fixed offsets and can therefore "opt-out" from the non-fragile ABI for ivars.

However, if an NSObject subclass has fixed offsets, then so must the subclasses of that subclass, so this allows us to optimize instances of subclasses of subclasses that inherit from NSObject and so on.

To determine this, we need to find that the compiler can see the implementation of each intermediate class, as that means it is statically linked.

Fixes: #81369

### CODE REVIEW

**Comment by AtariDreams** [comment]
@compnerd Thoughts on this?

**Comment by AtariDreams** [comment]
@alexander-shaposhnikov Thoughts?

**Comment by AtariDreams** [comment]
@jroelofs Could you please take a look when you get the chance?

**Comment by AtariDreams** [comment]
@rjmccall Do you think this is acceptable?

**Review by rjmccall — CHANGES_REQUESTED** [review]
Okay, your description of the change here is very misleading; it doesn't even mention that this is specific to having an `@implementation` for every intermediate class.

Your test seems completely inadequate — you need to test that we're doing the static layout properly, and you need to test with ivars coming from a variety of different places, including:
- ivars declared in the `@interface`
- ivars declared in the `@implementation`
- ivars synthesized explicitly in the `@implementation`
- vars synthesized implicitly in the `@implementation`

**Comment by AtariDreams** [comment]
> Okay, your description of the change here is very misleading; it doesn't even mention that this is specific to having an `@implementation` for every intermediate class.
> 
> Your test seems completely inadequate — you need to test that we're doing the static layout properly, and you need to test with ivars coming from a variety of different places, including:
> 
> * ivars declared in the `@interface`
> * ivars declared in the `@implementation`
> * ivars synthesized explicitly in the `@implementation`
> * vars synthesized implicitly in the `@implementation`

So this is a way to check for static linked; if the compiler can see the @implementation, right?

Only then can we "opt-out" of the non-fragile ABI like NSObject and its subclasses do, right?

**Comment by AtariDreams** [comment]
> test seems completely inadequate — you need to test that we're doing the static layout properly, and you need to test with ivars coming from a variety of different places, including:
> 
> * ivars declared in the `@interface`
> * ivars declared in the `@implementation`
> * ivars synthesized explicitly in the `@implementation`
> * vars synthesized implicitly in the `@implementation`

I think it is the only way, otherwise it would have to be done at link time. Can we do it at link time? probably another PR would have to deal with that.

**Comment by AtariDreams** [comment]
Layout before and after my change:
Screenshot 2024-02-10 at 6 55 53 PM

**Comment by AtariDreams** [comment]
Sorry, that was codegen, I meant 
image

**Comment by AtariDreams** [comment]
@rjmccall Are these tests adequate?

**Comment by rjmccall** [comment]
I talked this over with Mike Ash.  I was concerned that Objective-C might allow dynamic class changes that would burn our ability to do this optimization, such as adding ivars during `+initialize`.  The existing optimization would work even with a relatively lax runtime as long as new ivars always followed existing ivars.  Fortunately, Mike assures me that this is impossible: the layout of a class is fixed before it's ever exposed to user code, including the calls to `+load` and `+initialize`.  The runtime only allows ivars to be added to a class before the class is registered with the runtime (i.e. never for compiler-emitted classes).  It is possible to replace the superclass dynamically (although this is strongly discouraged), but class layout has already happened by any point where this is possible, so there is an effective requirement that the new superclass's class layout matches the old one's.  In short, we're comfortable saying that there's an ABI rule that class layout for compiled classes is solely dependent on information expressed in the language, at least as far as the class's subobject is concerned (that is, we can never rule out subclasses, including dynamic subclasses; but that doesn't interfere with the static optimization here).  So this optimization is indeed possible.

The compiler is not integrated with the linker and cannot see `@implementation`s for classes defined outside the current translation unit.  So the approach in this patch only helps in the case that you have the primary `@implementation` for multiple classes in a single translation unit.  That's fairly uncommon in my experience; ObjC programmers tend to define different classes in different `.m` files.  Still, it probably does happen, especially with code generation, and we ought to be able to optimize it.

We should stress to programmers that they should not change their code by defining everything in a single `.m` file just to get this optimization.  It is not likely to be worthwhile.

In order to optimize this across translation units, we'd need a version of it that worked with LTO.  It should be very feasible to do this with summaries in ThinLTO.  That seems like it'd be a huge win; it would have three effects that I can see:

- We'd be able to pre-slide offset variables in subclasses according to their superclasses in the same image.  This can significantly optimize load times by reducing the amount of memory that has to be dirtied.  However, I think the Mach-O linker already does this, so it's probably not a big deal to do in LTO.
- We'd be able to optimize ivar accesses to use static offsets even when the superclass is defined in a different TU.  This seems like the biggest win.
- We'd be able to optimize ivar accesses to use static offsets even for accesses outside of the main `@implementation`.  This would be a huge win except that it's pretty unidiomatic in Objective-C, which generally strongly encourages the use of properties.

I'll take another look at your patch.  As Aaron said, there is no need to ping me so quickly, especially on the weekend.

**Comment by AtariDreams** [comment]
> I talked this over with Mike Ash. I was concerned that Objective-C might allow dynamic class changes that would burn our ability to do this optimization, such as adding ivars during `+initialize`. The existing optimization would work even with a relatively lax runtime as long as new ivars always followed existing ivars. Fortunately, Mike assures me that this is impossible: the layout of a class is fixed before it's ever exposed to user code, including the calls to `+load` and `+initialize`. The runtime only allows ivars to be added to a class before the class is registered with the runtime (i.e. never for compiler-emitted classes). It is possible to replace the superclass dynamically (although this is strongly discouraged), but class layout has already happened by any point where this is possible, so there is an effective requirement that the new superclass's class layout matches the old one's. In short, we're comfortable saying that there's an ABI rule that class layout for compiled classes is solely dependent on information expressed in the language, at least as far as the class's subobject is concerned (that is, we can never rule out subclasses, including dynamic subclasses; but that doesn't interfere with the static optimization here). So this optimization is indeed possible.
> 
> The compiler is not integrated with the linker and cannot see `@implementation`s for classes defined outside the current translation unit. So the approach in this patch only helps in the case that you have the primary `@implementation` for multiple classes in a single translation unit. That's fairly uncommon in my experience; ObjC programmers tend to define different classes in different `.m` files. Still, it probably does happen, especially with code generation, and we ought to be able to optimize it.
> 
> We should stress to programmers that they should not change their code by defining everything in a single `.m` file just to get this optimization. It is not likely to be worthwhile.
> 
> In order to optimize this across translation units, we'd need a version of it that worked with LTO. It should be very feasible to do this with summaries in ThinLTO. That seems like it'd be a huge win; it would have three effects that I can see:
> 
> * We'd be able to pre-slide offset variables in subclasses according to their superclasses in the same image.  This can significantly optimize load times by reducing the amount of memory that has to be dirtied.  However, I think the Mach-O linker already does this, so it's probably not a big deal to do in LTO.
> * We'd be able to optimize ivar accesses to use static offsets even when the superclass is defined in a different TU.  This seems like the biggest win.
> * We'd be able to optimize ivar accesses to use static offsets even for accesses outside of the main `@implementation`.  This would be a huge win except that it's pretty unidiomatic in Objective-C, which generally strongly encourages the use of properties.
> 
> I'll take another look at your patch. As Aaron said, there is no need to ping me so quickly, especially on the weekend.

Alright. I think the LTO should be done as a follow-up since it will likely take some time and multiple reviews to get right.

**Review by rjmccall — CHANGES_REQUESTED** [review]
> @rjmccall Are these tests adequate?

No.  You need to test the behavior of ivar accesses in the subclass when there is an intermediate superclass that has ivars declared in all the ways I identified.  Your test has many tests for ivar accesses in immediate subclasses of `NSObject`, which is an optimization that we've already been doing for several years.  It only tests indirect subclasses in one configuration, which is when the intermediate superclass has all its ivars declared in the `@interface`.

**Comment by rjmccall** [comment]
> Alright. I think the LTO should be done as a follow-up PR since it will likely take some time and multiple reviews to get right.

Yes, that's a fine plan.  I expect that that'll be a significant amount of work.

**Comment by AtariDreams** [comment]
Ping

**Review by rjmccall — COMMENTED** [review]
Thanks, the breadth of tests looks good.  Please improve the actual testing, though — CHECK lines are testing the IR output file at a whole, so now that there are multiple tests in this file, we really need to make these tests more specific.

- First, please add CHECK-LABEL lines for the functions you're trying to test so that we know that the checks are actually applying to the code in that function.
- Second, I know the existing test is doing this CHECK-NOT thing, but it should really always have been testing that the ivar access actually uses the right offset; please do that, both for your new tests and the existing one.
- Finally, please make sure you test at least one access for each of the ivars.  Putting accesses in separate methods will help with ensuring that you're testing each specifically.

As a procedural note, I was going to review this last week after you responded to the review, but then I saw that you kept uploading new versions of the patch, so I held off.  I certainly understand needing to update the patch a few times; I do that myself all the time.  Just make sure you give me a heads up that the patch has been updated and is actually ready for review: either hold your response until your patch is ready, or make a short comment like "Sorry about that, patch is ready for review now." when the updates are done.

File: clang/test/CodeGenObjC/constant-non-fragile-ivar-offset.m
Comment (rjmccall) [inline clang/test/CodeGenObjC/constant-non-fragile-ivar-offset.m:4]: Please add a test for the access to this.  It'll be in a synthesized method.

File: clang/test/CodeGenObjC/constant-non-fragile-ivar-offset.m
Comment (rjmccall) [inline clang/test/CodeGenObjC/constant-non-fragile-ivar-offset.m:41]: Please add CHECK-LABEL lines for the `define` lines for each of the method definitions containing these ivar accesses.  You can look at test/CodeGenObjC/direct-methods.m for the pattern I'm looking for here.

**Comment by AtariDreams** [comment]
@rjmccall Done!

**Comment by AtariDreams** [comment]
Refactored for code consistency.

**Comment by AtariDreams** [comment]
Wait, @rjmccall   // Assume that anything declared with a forward declaration and no
  // @interface subclasses NSObject.
  
  Any reason why analysis can assume that but not codegen? Just curious.

**Review by rjmccall — APPROVED** [review]
Thanks, that looks great.

**Comment by AtariDreams** [comment]
@rjmccall Ready to merge!

**Comment by AtariDreams** [comment]
@rjmccall I do not have merge permissions or whatever it is called.

**Comment by AtariDreams** [comment]
@rjmccall ping

**Comment by AtariDreams** [comment]
@JOE1994 Thoughts on this?

**Comment by JOE1994** [comment]
> @JOE1994 Thoughts on this?

I'm not an ideal reviewer for this as I'm unfamiliar with ObjC side of Clang,
but I'll come back to leave a review and to help merge this if this doesn't get merged by the end of the day :)

**Comment by AtariDreams** [comment]
> > @JOE1994 Thoughts on this?
> 
> 
> 
> I'm not an ideal reviewer for this as I'm unfamiliar with ObjC side of Clang,
> 
> but I'll come back to leave a review and to help merge this if this doesn't get merged by the end of today :)

Alright. Thank you so much!

**Review by JOE1994 — APPROVED** [review]
LGTM

**Comment by nocchijiang** [comment]
@AZero13 Do you have any news or plans regarding LTO support? We have several projects with substantial existing Objective-C code, and the performance benefits of LTO support could be significant. I would be interested in taking on a portion or the entirety of this work if it is not currently on your priority list.

**Comment by AZero13** [comment]
> @AZero13 Do you have any news or plans regarding LTO support? We have several projects with substantial existing Objective-C code, and the performance benefits of LTO support could be significant. I would be interested in taking on a portion or the entirety of this work if it is not currently on your priority list.

If you can do it, go for it!

**Pull request merged**
